### DISCUSSION TITLE
Compilation errors when using erlang 25

### DISCUSSION THREAD

**Discussion by tomyouyou** [rabbitmq/rabbitmq-server#4867] [Other] [answered]
Hi, when I compiled rabbitmq-server-3.10.1 with erlang 25, the following error occurred. 
How to solve the compilation error? Thanks.

```
15:14:02.392 [error] beam/beam_load.c(551): Error loading function 'Elixir.Hex.Version':'allow_pre_available?'/0: op put_tuple u x:
  please re-compile this module with an Erlang/OTP 25 compiler



15:14:02.392 [error] Loading of /root/rpmbuild/SOURCES/rabbitmq-server-3.10.1/deps/.mix/archives/hex-1.0.1/hex-1.0.1/ebin/Elixir.Hex.Version.beam failed: :badfile


15:14:02.393 [notice] Application hex exited: exited in: Hex.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function Hex.Version.start/0 is undefined (module Hex.Version is not available)
            (hex 1.0.1) Hex.Version.start()
            (hex 1.0.1) lib/hex/application.ex:12: Hex.Application.start/2
            (kernel 8.4) application_master.erl:293: :application_master.start_it_old/4

15:14:02.404 [notice] Application inets exited: :stopped

15:14:02.405 [notice] Application ssl exited: :stopped

15:14:02.405 [notice] Application public_key exited: :stopped

15:14:02.405 [notice] Application asn1 exited: :stopped

15:14:02.405 [notice] Application crypto exited: :stopped
Could not start Hex. Try fetching a new version with "mix local.hex" or uninstalling it with "mix archive.uninstall hex.ez"
** (MatchError) no match of right hand side value: {:error, {:hex, {:bad_return, {{Hex.Application, :start, [:normal, []]}, {:EXIT, {:undef, [{Hex.Version, :start, [], []}, {Hex.Application, :start, 2, [file: 'lib/hex/application.ex', line: 12]}, {:application_master, :start_it_old, 4, [file: 'application_master.erl', line: 293]}]}}}}}}
    (hex 1.0.1) lib/hex.ex:5: Hex.start/0
    (mix 1.13.4) lib/mix/hex.ex:59: Mix.Hex.start/0
    (mix 1.13.4) lib/mix/dep/loader.ex:203: Mix.Dep.Loader.with_scm_and_app/4
    (mix 1.13.4) lib/mix/dep/loader.ex:156: Mix.Dep.Loader.to_dep/3
    (elixir 1.13.4) lib/enum.ex:1593: Enum."-map/2-lists^map/1-0-"/2
    (mix 1.13.4) lib/mix/dep/loader.ex:373: Mix.Dep.Loader.mix_children/2
    (mix 1.13.4) lib/mix/dep/loader.ex:18: Mix.Dep.Loader.children/0
    (mix 1.13.4) lib/mix/dep/converger.ex:57: Mix.Dep.Converger.all/4
make[2]: *** [Makefile:108: escript/rabbitmqctl] Error 1
make[2]: Leaving directory '/root/rpmbuild/SOURCES/rabbitmq-server-3.10.1/deps/rabbitmq_cli'
make[1]: *** [../../erlang.mk:4512: deps] Error 2
make[1]: Leaving directory '/root/rpmbuild/SOURCES/rabbitmq-server-3.10.1/deps/rabbit'
make: *** [erlang.mk:4512: deps] Error 2
```

**Comment by michaelklishin** [comment]
Please use Discussions for questions.

**Comment by michaelklishin** [comment, marked as answer]
Erlang 25 no longer can load modules compiled on Erlang 21 or older. Elixir 1.10.x was compiled on 21.
We had to switch our pipelines to use Elixir 1.13.x even though we do not really have any reasons to depend on it, and then compile from source to never depend on what version Elixir maintainers pick at any point in the future.

**Comment by tomyouyou** [comment]
@michaelklishin  
thank you.
I recompiled hex module on erlang 25 and replaced all the files in the rabbitmq-server-3.10.1/deps/.mix/archives/hex-1.0.1/hex-1.0.1/ebin directory.
Now the compilation  is ok.

**Reply by loicknuchel** [reply]
Hi, I'm a total newcomer, any tip on how to recompile with erlang 25?

**Reply by michaelklishin** [reply]
The answer is always "do not compile from source". If you really have to do it…

You need to make sure that the Elixir version you use in the environment was built with Erlang 23+,
maybe even 25. Usually `kiex` is the easiest option to get there, with package managers you don't control the complication toolchain. I see no reason to use Elixir versions older than 1.13 if you can target Erlang exclusively.

Then just compile RabbitMQ as usual using Erlang 25.

**Comment by lcr0815** [comment]
my question is:

when you say recompile so and so with v25, does that mean one has to down load some version of the software that has to be **recompiled**. I am using  the downloaded windows installer for both elixir and erlang so whatever they download is what is there.

how do I recompile in erl 25 from the download....

:(

**Reply by michaelklishin** [reply]
Installing latest Erlang and Elixir should be enough. Older Elixir versions were compiled on Erlang 21 which Erlang 25 can no longer load.
